monitoring/ping.go

```diff
@@ -0,0 +1,295 @@
+type pingChecker struct {
+	self           serf.Member
+	serfClient     agent.SerfClient
+	serfRPCAddr    string
+	serfMemberName string
+	latencyStats   ttlmap.TTLMap
+	mux            sync.Mutex
+	logger         log.Entry
+}
+
+type PingCheckerConfig struct {
+	SerfRPCAddr    string
+	SerfMemberName string
+	NewSerfClient  agent.NewSerfClientFunc
+}
+
+func (c *PingCheckerConfig) CheckAndSetDefaults() error {
+	if c.SerfRPCAddr == "" {
+		return trace.BadParameter("serf rpc address can't be empty")
+	}
+	if c.SerfMemberName == "" {
+		return trace.BadParameter("serf member name can't be empty")
+	}
+	if c.NewSerfClient == nil {
+		c.NewSerfClient = agent.NewSerfClient
+	}
+	return nil
+}
+
+// NewPingChecker returns a checker that verifies accessibility of nodes in the cluster by exchanging ping requests
+func NewPingChecker(conf PingCheckerConfig) (c health.Checker, err error) {
+	err = conf.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	latencyTTLMap, err := ttlmap.New(latencyStatsCapacity)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	logger := log.WithFields(log.Fields{trace.Component: "ping"})
+	logger.Debugf("using Serf IP: %v", conf.SerfRPCAddr)
+	logger.Debugf("using Serf Name: %v", conf.SerfMemberName)
+
+	client, err := conf.NewSerfClient(serf.Config{
+		Addr: conf.SerfRPCAddr,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	// finding what is the current node
+	var self serf.Member
+	for _, node := range nodes {
+		if node.Name == conf.SerfMemberName {
+			self = node
+			break // self node found, breaking out of the for loop
+		}
+	}
+	if self.Name == "" {
+		return nil, trace.NotFound("failed to find Serf member with name %s", conf.SerfMemberName)
+	}
+
+	return &pingChecker{
+		self:           self,
+		serfClient:     client,
+		serfRPCAddr:    conf.SerfRPCAddr,
+		serfMemberName: conf.SerfMemberName,
+		latencyStats:   *latencyTTLMap,
+		logger:         *logger,
+	}, nil
+}
+
+// Name returns the checker name
+// Implements health.Checker
+func (c *pingChecker) Name() string {
+	return pingCheckerID
+}
+
+// Check verifies that all nodes' ping with Master Nodes is lower than the
+// desired threshold
+// Implements health.Checker
+func (c *pingChecker) Check(ctx context.Context, r health.Reporter) {
+	err := c.check(ctx, r)
+
+	if err != nil {
+		c.logger.Error(err.Error())
+		r.Add(NewProbeFromErr(c.Name(), "", err))
+		return
+	}
+	r.Add(&pb.Probe{Checker: c.Name(), Status: pb.Probe_Running})
+}
+
+// check runs the actual system status verification code and returns an error
+// in case issues arise in the process
+func (c *pingChecker) check(ctx context.Context, r health.Reporter) (err error) {
+
+	client := c.serfClient
+
+	// retrieve other nodes using Serf members
+	nodes, err := client.Members()
```

Comment:
I'm curious - how does this handle removed nodes? I.e. when you shrink a cluster, serf member stays in the "left" state so will it be returned here? And we'll keep trying to measure latency to it? I think we should test this scenario and maybe filter out all non-active members here.